pycalphad/core/workspace.py

```diff
@@ -0,0 +1,462 @@
+            value = instantiate_models(obj.database, obj.components, obj.phases, model=value, parameters=obj.parameters)
+            super().__set__(obj, value)
+        except AttributeError:
+            super().__set__(obj, None)
+
+    def on_dependency_update(self, obj, updated_attribute, old_val, new_val):
+        self.__set__(obj, self.default_factory(obj))
+
+class PRFField(TypedField):
+    def __init__(self, dependsOn=None):
+        def make_prf(obj):
+            try:
+                prf = PhaseRecordFactory(obj.database, obj.components, obj.conditions,
+                                         obj.models.unwrap() if hasattr(obj.models, 'unwrap') else obj.models,
+                                         parameters=obj.parameters)
+                return prf
+            except AttributeError:
+                return None
+        super().__init__(default_factory=make_prf, dependsOn=dependsOn)
+
+    def on_dependency_update(self, obj, updated_attribute, old_val, new_val):
+        self.__set__(obj, self.default_factory(obj))
+
+class SolverField(TypedField):
+    def on_dependency_update(self, obj, updated_attribute, old_val, new_val):
+        self.__set__(obj, self.default_factory(obj))
+
+class EquilibriumCalculationField(TypedField):
+    def __get__(self, obj, objtype=None):
+        if (not hasattr(obj, self.private_name)) or (getattr(obj, self.private_name) is None):
+            try:
+                default_value = obj.recompute()
+            except AttributeError:
+                default_value = None
+            setattr(obj, self.private_name, default_value)
+        return getattr(obj, self.private_name)
+
+    def on_dependency_update(self, obj, updated_attribute, old_val, new_val):
+        self.__set__(obj, None)
+
+
+# Defined to allow type checking for Model or its subclasses
+ModelType = TypeVar('ModelType', bound=Model)
+
+class Workspace:
+    _callbacks = defaultdict(lambda: [])
+    database: Database = TypedField(lambda _: None)
+    components: SpeciesList = ComponentsField(dependsOn=['database'])
+    phases: PhaseList = PhasesField(dependsOn=['database', 'components'])
+    conditions: Conditions = ConditionsField(lambda wks: Conditions(wks))
+    verbose: bool = TypedField(lambda _: False)
+    models: Mapping[PhaseName, ModelType] = ModelsField(dependsOn=['phases', 'parameters'])
+    parameters: SumType([NoneType, Dict]) = DictField(lambda _: OrderedDict())
+    renderer: Renderer = TypedField(lambda wks: DEFAULT_PLOT_RENDERER(wks))
+    phase_record_factory: Optional[PhaseRecordFactory] = PRFField(dependsOn=['phases', 'conditions', 'models', 'parameters'])
+    calc_opts: SumType([NoneType, Dict]) = DictField(lambda _: OrderedDict())
+    solver: SolverBase = SolverField(lambda obj: Solver(verbose=obj.verbose), dependsOn=['verbose'])
+    eq: Optional[LightDataset] = EquilibriumCalculationField(dependsOn=['phase_record_factory', 'calc_opts', 'solver'])
```

Comment:
Not sure where it lives (in `Workspace` or in `TypedField` or both), but it would be nice to document that `eq` is set by a callback from `TypedField`

Reply:
Comment added in `Workspace`